#726(2) documentation update reviews level 1 + (#909)
openmsupply/application-manager-web-app: docs/internal/Level-1+-Reviews-and-Changes-Requested.md

# Level 1+ Reviews and Changes Requested

> The [Level 1 Reviews](Level-1-Review-And-LOQ.md) explains how reviews with only 1 level works. Which is a subset of how a review with more levels (per stage) should work.

Reviewer: ```suggestion
The [Level 1 Reviews](Level-1-Review-And-LOQ.md) page explains how reviews with only one level work, which is a subset of how a review with more levels (per stage) should work.
```

The aim of this page is to outline the structure changes and flows:

> ## Consolidation in agreement

Reviewer: Be good if the description part of these in the list had matching tense. So either:
- Reviewer (Level 1) - reviewing an application
- Reviewer (Level 1) - submitting ~~`LOQ (List of Questions)`~~ `Conform` or `Non Conform` advise to upper level reviewer

OR:

- Reviewer (Level 1) - review an application
- Reviewer (Level 1) - submit ~~`LOQ (List of Questions)`~~ `Conform` or `Non Conform` advise to upper level reviewer

But not a mixture of both

Reviewer: Ah, I see on the Leve1 review page you've used the "ing" tense in a similar list, so probably best to go with that.

Author: Ok. Will add `ing` to where is missing

- Reviewer (Level 1) - reviewing an application

> - Reviewer (Level 1) - submit ~~`LOQ (List of Questions)`~~ `Conform` or `Non Conform` advise to upper level reviewer

Reviewer: ```suggestion
- Reviewer (Level 1) - submit ~~`LOQ (List of Questions)`~~ `Conform` or `Non Conform` to advise higher level reviewer
```
Also -- is the ~~strikethrough~~ helpful?

Author: Yea,no. I'll remove the LOQ part

- Reviewer (Level 2) - reviewing the lower level review and `agree` with all advised decisions
- Reviewer (Level 2) - submit a review decision - if user has `is_last_level` will send back to Applicant when `LOQ (List of Questions)` or change outcome and stage when submitting `Non Conform` or `Conform`. If consolidation not `is_last_level` submission of review goes to upper level reviewer (e.g. Level 3) and the same process repeats for Level 2 and Level 3 reveiwers.

## Consolidation in disagreement (Changes-requested)

- Reviewer (Level 1) - reviewing an application

> - Reviewer (Level 1) - submit ~~`LOQ (List of Questions)`~~ `Conform` or `Non Conform` advise to upper level reviewer

Reviewer: Same as above

- Reviewer (Level 2) - consolidate lower level review(s) and `disagree` with some advised decisions
- Reviewer (Level 2) - submit `Changes Requested` back to reviewer (e.g. Level 1)
- Reviewer (Level 1) - makes changes in response to `Changes Requested` and re-submit to upper level reviewer.

> ### Schema changes

Reviewer: Wouldn't this be more correctly called "Structure changes"?

Author: Good point. Done

- **TODO**: Not really sure exact the current use of `latestOriginalReviewResponse` and `previousOriginalReviewResponse` so will wait for feedback to complete this one....

#### Review

- The progress for a first-review (if no changes-requested) is done only with (`reviewProgress`)
- If some changes requested:
  - used both `reviewProgress` and `changeRequestsProgress` to track review progress
  - each element `state` has `isChangeRequest` and `isChanged` (if one of the change-requested) to track if changes are done
- On the review, similarly to consolidation each element `state` has:
  - the decision done in this current **review** in `thisReviewLatestResponse`
  - the decision done in a previous **review** (if any) in `thisReviewPreviousResponse` to track changes between this and a previous review (of the same user)
- To find the consolidator change-requested to this reviewer:
  - if it is stored inside `thisReviewPreviousResponse` (when review is in DRAFT) or `thisReviewLatestResponse` (if review in CHANGES_REQUESTED hasn't started)

## Submitting back to Applicant

> When a upper level reviewer disagrees with some decision on lower level it needs to be changed by lower level on a new review.

Reviewer: ```suggestion
When an upper level reviewer disagrees with some decision on a lower level it needs to be changed by the lower level reviewer on a new review.
```

> The consolidation **needs to happen again and be in agreement** so it can be send to Applicant (with Reviewer Level 1 comments) in the form of `LOQ (List of Questions)`.

Reviewer: ```suggestion
The consolidation **needs to happen again and be in agreement** so it can be sent to Applicant (with Reviewer Level 1 comments) in the form of `LOQ (List of Questions)`.
```

- GraphQL mutation to create review responses - linked to another review or to application response
- GraphQL mutation to update `review_response` and `review_decison`
- Validates submission of review and give options to submmit based on `is_last_level` and if review is level 1+
- GraphQL mutation to request submission of review

Similart to what was in previous [Level 1 Reviews](Level-1-Review-And-LOQ.md), but considering review of review.

## Validation of Submission

### Review

First review validation in more details on [Level 1 Review and LOQ](Level-1-Review-And-LOQ.md).

When **review** is updated as a result of `Changes Requested` by a Level 1+ reviewer, it can only be submitted if general validation passes AND all questions with change requests have been changed.

Logic to check validation of a **changes-requested review** is:

- `thisReviewLastestResponse` is different to `thisReviewPreviousResponse` that has some associated `reviewResponse.reviewResponseLinkId` with decision of `disagree`. This is more easily validated with fields `isChangeRequest` and `isChanged` in the 'state' of each element.

### Consolidation

Logic to check validation is:

> - On `disagreeement` it will configure a **changes-quested** decision and submitted back to other reviewers

Reviewer: ```suggestion
- On `disagreement` it will configure a **changes-requested** decision and be submitted back to other reviewers
```
Also -- is "configure" the right term here? Could it be "create"?

Author: Yes, if that's clearer. I'll change now
